docs/ALU Design.md

```diff
@@ -0,0 +1,758 @@
+  - `__llvm_bitcast_i8_to_i32 -> i32`
+  - `__llvm_bitcast_i8_to_i64 -> i64`
+  - `__llvm_bitcast_i8_to_i128 -> i128`
+  - `__llvm_bitcast_i16_to_bool -> bool`
+  - `__llvm_bitcast_i16_to_i8 -> i8`
+  - `__llvm_bitcast_i16_to_i16 -> i16`
+  - `__llvm_bitcast_i16_to_i32 -> i32`
+  - `__llvm_bitcast_i16_to_i64 -> i64`
+  - `__llvm_bitcast_i16_to_i128 -> i128`
+  - `__llvm_bitcast_i32_to_bool -> bool`
+  - `__llvm_bitcast_i32_to_i8 -> i8`
+  - `__llvm_bitcast_i32_to_i16 -> i16`
+  - `__llvm_bitcast_i32_to_i32 -> i32`
+  - `__llvm_bitcast_i32_to_i64 -> i64`
+  - `__llvm_bitcast_i32_to_i128 -> i128`
+  - `__llvm_bitcast_i64_to_bool -> bool`
+  - `__llvm_bitcast_i64_to_i8 -> i8`
+  - `__llvm_bitcast_i64_to_i16 -> i16`
+  - `__llvm_bitcast_i64_to_i32 -> i32`
+  - `__llvm_bitcast_i64_to_i64 -> i64`
+  - `__llvm_bitcast_i64_to_i128 -> i128`
+  - `__llvm_bitcast_i128_to_bool -> bool`
+  - `__llvm_bitcast_i128_to_i8 -> i8`
+  - `__llvm_bitcast_i128_to_i16 -> i16`
+  - `__llvm_bitcast_i128_to_i32 -> i32`
+  - `__llvm_bitcast_i128_to_i64 -> i64`
+  - `__llvm_bitcast_i128_to_i128 -> i128`
+  - `__llvm_bitcast_pbool_to_pbool -> pbool`
+  - `__llvm_bitcast_pbool_to_pi8 -> pi8`
+  - `__llvm_bitcast_pbool_to_pi16 -> pi16`
+  - `__llvm_bitcast_pbool_to_pi32 -> pi32`
+  - `__llvm_bitcast_pbool_to_pi64 -> pi64`
+  - `__llvm_bitcast_pbool_to_pi128 -> pi128`
+  - `__llvm_bitcast_pi8_to_pbool -> pbool`
+  - `__llvm_bitcast_pi8_to_pi8 -> pi8`
+  - `__llvm_bitcast_pi8_to_pi16 -> pi16`
+  - `__llvm_bitcast_pi8_to_pi32 -> pi32`
+  - `__llvm_bitcast_pi8_to_pi64 -> pi64`
+  - `__llvm_bitcast_pi8_to_pi128 -> pi128`
+  - `__llvm_bitcast_pi16_to_pbool -> pbool`
+  - `__llvm_bitcast_pi16_to_pi8 -> pi8`
+  - `__llvm_bitcast_pi16_to_pi16 -> pi16`
+  - `__llvm_bitcast_pi16_to_pi32 -> pi32`
+  - `__llvm_bitcast_pi16_to_pi64 -> pi64`
+  - `__llvm_bitcast_pi16_to_pi128 -> pi128`
+  - `__llvm_bitcast_pi32_to_pbool -> pbool`
+  - `__llvm_bitcast_pi32_to_pi8 -> pi8`
+  - `__llvm_bitcast_pi32_to_pi16 -> pi16`
+  - `__llvm_bitcast_pi32_to_pi32 -> pi32`
+  - `__llvm_bitcast_pi32_to_pi64 -> pi64`
+  - `__llvm_bitcast_pi32_to_pi128 -> pi128`
+  - `__llvm_bitcast_pi64_to_pbool -> pbool`
+  - `__llvm_bitcast_pi64_to_pi8 -> pi8`
+  - `__llvm_bitcast_pi64_to_pi16 -> pi16`
+  - `__llvm_bitcast_pi64_to_pi32 -> pi32`
+  - `__llvm_bitcast_pi64_to_pi64 -> pi64`
+  - `__llvm_bitcast_pi64_to_pi128 -> pi128`
+  - `__llvm_bitcast_pi128_to_pbool -> pbool`
+  - `__llvm_bitcast_pi128_to_pi8 -> pi8`
+  - `__llvm_bitcast_pi128_to_pi16 -> pi16`
+  - `__llvm_bitcast_pi128_to_pi32 -> pi32`
+  - `__llvm_bitcast_pi128_to_pi64 -> pi64`
+  - `__llvm_bitcast_pi128_to_pi128 -> pi128`
+- [`icmp`](https://llvm.org/docs/LangRef.html#icmp-instruction):
+  - `icmp` accepts three arguments. The first is a comparison condition. The two others are operands
+    of the same type. The condition is defined as an enum consisting of these values: `eq`, `ne`,
+    `ugt`, `uge`, `ult`, `ule`, `sgt`, `sge`, `slt`, `sle`.
```

Comment:
You should specify whether we are taking the enum at runtime as an argument or whether the compiler needs to translate each comparison to a static call.

Reply:
I mean it would be more efficient at execution time to have all 60 possible combinations rather than having to dispatch on the enum argument, but by the same token we don't actually care that much about performance _now_. I leave this up to you, just please document what the decision is so we don't end up with the compiler's code generation and the polyfills themselves disagreeing.